File: src/leetcode/java/hard/TrappingRainWaterII_407.java
package leetcode.java.hard;

import java.util.PriorityQueue;

public class TrappingRainWaterII_407 {
  private int water;
  private boolean[][] visited;

  public int trapRainWater(int[][] heightMap) {
    if (heightMap.length == 0) return 0;
    int row = heightMap.length;
    int col = heightMap[0].length;
    water = 0;
    PriorityQueue<Cell> walls = new PriorityQueue<>();
    visited = new boolean[row][col];
    for (int i = 0; i < col; i++) {
      walls.offer(new Cell(0, i, heightMap[0][i]));
      visited[0][i] = true;
      walls.offer(new Cell(row - 1, i, heightMap[row - 1][i]));
      visited[row - 1][i] = true;
    }
    for (int i = 1; i < row - 1; i++) {
      walls.offer(new Cell(i, 0, heightMap[i][0]));
      visited[i][0] = true;
      walls.offer(new Cell(i, col - 1, heightMap[i][col - 1]));
      visited[i][col - 1] = true;
    }
    while (!walls.isEmpty()) {
      Cell min = walls.poll();
      visit(heightMap, min, walls);
    }
    return water;
  }

  private void visit(int[][] heightMap, Cell min, PriorityQueue<Cell> walls) {
    fill(heightMap, min.x + 1, min.y, min.h, walls);
    fill(heightMap, min.x - 1, min.y, min.h, walls);
    fill(heightMap, min.x, min.y + 1, min.h, walls);
    fill(heightMap, min.x, min.y - 1, min.h, walls);
  }

  private void fill(int[][] heightMap, int x, int y, int h, PriorityQueue<Cell> walls) {
    if (x < 0 || y < 0 || x >= heightMap.length || y >= heightMap[0].length || visited[x][y]) return;
    else if (heightMap[x][y] >= h) {
      walls.offer(new Cell(x, y, heightMap[x][y]));
      visited[x][y] = true;
      return;
    } else {
      water += h - heightMap[x][y];
      visited[x][y] = true;
      fill(heightMap, x + 1, y, h, walls);
      fill(heightMap, x - 1, y, h, walls);
      fill(heightMap, x, y + 1, h, walls);
      fill(heightMap, x, y - 1, h, walls);
    }
  }

  class Cell implements Comparable<Cell> {
    int x, y, h;
    Cell(int x, int y, int h) {
      this.x = x;
      this.y = y;
      this.h = h;
    }
    @Override
    public int compareTo(Cell o) {
      return h - o.h;
    }
  }
}

File: src/leetcode/python/TrappingRainWaterII_407.py
class TrappingRainWaterII_407(object):
    def trapRainWater(self, heightMap):
        """
        :type heightMap: List[List[int]]
        :rtype: int
        """
        m = len(heightMap)
        if m == 0:
            return 0
        n = len(heightMap[0])
        visited = [[False for i in range(n)] for j in range(m)]
        from queue import PriorityQueue
        q = PriorityQueue()
        for i in range(m):
            visited[i][0] = True
            q.put([heightMap[i][0], i, 0])
            visited[i][n-1] = True
            q.put([heightMap[i][n-1], i, n-1])
        for j in range(1, n-1):
            visited[0][j] = True
            q.put([heightMap[0][j], 0, j])
            visited[m-1][j] = True
            q.put([heightMap[m-1][j], m-1, j])
        water = 0
        while not q.empty():
            cell = q.get()
            for (i, j) in [(1,0), (-1,0), (0,1), (0,-1)]:
                x = cell[1] + i
                y = cell[2] + j
                if x in range(m) and y in range(n) and not visited[x][y]:
                    water += max(0, cell[0] - heightMap[x][y])
                    q.put([max(heightMap[x][y], cell[0]), x, y])
                    visited[x][y] = True
        return water
